File: transformer/src/main/java/com/daykm/p5executioner/Main.java
package com.daykm.p5executioner;

import com.daykm.p5executioner.json.JsonCombo;
import com.daykm.p5executioner.json.JsonPersonaDetail;
import com.daykm.p5executioner.json.JsonSkillDetail;
import com.daykm.p5executioner.json.JsonSpecialCombo;
import com.daykm.p5executioner.proto.Arcana;
import com.daykm.p5executioner.proto.ArcanaCombo;
import com.daykm.p5executioner.proto.DLCPersona;
import com.daykm.p5executioner.proto.Data;
import com.daykm.p5executioner.proto.Persona;
import com.daykm.p5executioner.proto.RareComboModifier;
import com.daykm.p5executioner.proto.Skill;
import com.daykm.p5executioner.proto.SpecialCombo;
import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;
import com.squareup.moshi.Types;
import java.io.File;
import java.lang.reflect.ParameterizedType;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import okio.Okio;
import okio.Source;

public class Main {

	private static final String OUTPUT = "gen/src/main/assets/";
	private static final String INPUT = "transformer/json/";

	public static void main(String[] args) throws Exception {
		System.out.println("Working Directory = " + System.getProperty("user.dir"));

		Moshi moshi = new Moshi.Builder().build();

		Data data = Data.newBuilder()
				.addAllArcanaCombos(createArcanaCombos(moshi))
				.addAllPersonas(createPersonas(moshi))
				.addAllSkills(createSkills(moshi))
				.addAllRareModifiers(createRareCombos(moshi))
				.addAllSpecialCombos(createSpecialCombos(moshi))
				.addAllDlcPersonae(createDlcPersonae(moshi))
				.build();

		Files.write(Paths.get(OUTPUT + "data.pb"), data.toByteArray());

		data = Data.parseFrom(Files.readAllBytes(Paths.get(OUTPUT + "data.pb")));

		System.out.print("finished");
	}

	public static List<DLCPersona> createDlcPersonae(Moshi moshi) throws Exception {
		JsonAdapter<List<List<String>>> adapter = moshi.adapter(Types.newParameterizedType(List.class,
				Types.newParameterizedType(List.class, String.class)));

		List<List<String>> serieses =
				adapter.fromJson(Okio.buffer(Okio.source(new File(INPUT + "dlcPersonae.json"))));

		List<DLCPersona> protoPersonae = new ArrayList<>(serieses.size());
		for (List<String> series : serieses) {
			protoPersonae.add(DLCPersona.newBuilder().addAllPersonaeInSeries(series).build());
		}
		return protoPersonae;
	}

	public static List<SpecialCombo> createSpecialCombos(Moshi moshi) throws Exception {

		JsonAdapter<List<JsonSpecialCombo>> adapter =
				moshi.adapter(Types.newParameterizedType(List.class, JsonSpecialCombo.class));

		List<JsonSpecialCombo> combos =
				adapter.fromJson(Okio.buffer(Okio.source(new File(INPUT + "specialCombos.json"))));

		List<SpecialCombo> protoCombos = new ArrayList<>(combos.size());

		for (JsonSpecialCombo combo : combos) {
			protoCombos.add(
					SpecialCombo.newBuilder().setResult(combo.result).addAllSources(combo.sources).build());
		}

		return protoCombos;
	}

	public static List<ArcanaCombo> createArcanaCombos(Moshi moshi) throws Exception {
		JsonAdapter<List<JsonCombo>> adapter =
				moshi.adapter(Types.newParameterizedType(List.class, JsonCombo.class));

		List<JsonCombo> json =
				adapter.fromJson(Okio.buffer(Okio.source(new File(INPUT + "arcanaCombos.json"))));

		List<ArcanaCombo> protoCombos = new ArrayList<>(json.size());

		for (JsonCombo combo : json) {
			protoCombos.add(ArcanaCombo.newBuilder()
					.setFirst(Arcana.valueOf(combo.source[0].toUpperCase()))
					.setSecond(Arcana.valueOf(combo.source[1].toUpperCase()))
					.setResult(Arcana.valueOf(combo.result.toUpperCase()))
					.build());
		}

		return protoCombos;
	}

	public static List<Skill> createSkills(Moshi moshi) throws Exception {

		ParameterizedType type =
				Types.newParameterizedType(Map.class, String.class, JsonSkillDetail.class);

		JsonAdapter<Map<String, JsonSkillDetail>> adapter = moshi.adapter(type);

		Source source = Okio.source(new File(INPUT + "skills.json"));
		Map<String, JsonSkillDetail> skills = adapter.fromJson(Okio.buffer(source));

		List<Skill> protoSkills = new ArrayList<>(skills.size());

		for (String key : skills.keySet()) {
			JsonSkillDetail detail = skills.get(key);

			Skill.PersonaAndLevel.Builder builder = Skill.PersonaAndLevel.newBuilder();
			for (String persona : detail.personas.keySet()) {
				builder.setPersona(persona).setLevel(detail.personas.get(persona));
			}
			protoSkills.add(Skill.newBuilder()
					.addAllTalks(detail.talks != null ? detail.talks : new ArrayList<String>(0))
					.setCost(detail.cost != null ? detail.cost : 0)
					.setName(key)
					.addPersonaeWhoLearn(builder)
					.build());
		}
		return protoSkills;
	}

	public static List<RareComboModifier> createRareCombos(Moshi moshi) throws Exception {

		ParameterizedType type = Types.newParameterizedType(Map.class, String.class,
				Types.newParameterizedType(List.class, Integer.class));

		JsonAdapter<Map<String, List<Integer>>> adapter = moshi.adapter(type);

		Source source = Okio.source(new File(INPUT + "rareCombos.json"));
		Map<String, List<Integer>> rareCombos = adapter.fromJson(Okio.buffer(source));

		List<RareComboModifier> protoModifiers = new ArrayList<>(rareCombos.size());

		for (String key : rareCombos.keySet()) {
			protoModifiers.add(RareComboModifier.newBuilder()
					.setArcana(Arcana.valueOf(key.toUpperCase()))
					.addAllModifiers(rareCombos.get(key))
					.build());
		}

		return protoModifiers;
	}

	public static List<Persona> createPersonas(Moshi moshi) throws Exception {
		ParameterizedType type =
				Types.newParameterizedType(Map.class, String.class, JsonPersonaDetail.class);

		JsonAdapter<Map<String, JsonPersonaDetail>> adapter = moshi.adapter(type);

		Source source = Okio.source(new File(INPUT + "personae.json"));

		Map<String, JsonPersonaDetail> personae = adapter.fromJson(Okio.buffer(source));

		List<Persona> protoPersonas = new ArrayList<>(personae.size());

		for (String key : personae.keySet()) {
			JsonPersonaDetail detail = personae.get(key);
			Persona.Builder builder = Persona.newBuilder();

			List<Persona.Affinities.AffinityOption> options = new ArrayList<>();

			for (String affinity : detail.elems) {
				switch (affinity) {
					case "ab":
						options.add(Persona.Affinities.AffinityOption.ABSORB);
						break;
					case "-":
						options.add(Persona.Affinities.AffinityOption.NONE);
						break;
					case "wk":
						options.add(Persona.Affinities.AffinityOption.WEAK);
						break;
					case "rs":
						options.add(Persona.Affinities.AffinityOption.RESIST);
						break;
					case "rp":
						options.add(Persona.Affinities.AffinityOption.REPEL);
						break;
					case "nu":
						options.add(Persona.Affinities.AffinityOption.NULL);
						break;
					default:
						options.add(Persona.Affinities.AffinityOption.UNRECOGNIZED);
				}
			}

			builder.setAffinities(Persona.Affinities.newBuilder()
					.setPhysical(options.get(0))
					.setGun(options.get(1))
					.setFire(options.get(2))
					.setIce(options.get(3))
					.setElectric(options.get(4))
					.setWind(options.get(5))
					.setPsychic(options.get(6))
					.setNuclear(options.get(7))
					.setBless(options.get(8))
					.setCurse(options.get(9)));

			builder.setStats(Persona.Stats.newBuilder()
					.setStrength(detail.stats[0])
					.setMagic(detail.stats[1])
					.setEndurance(detail.stats[2])
					.setAgility(detail.stats[3])
					.setLuck(detail.stats[4]));

			for (String skill : detail.skills.keySet()) {
				builder.addSkills(
						Persona.LearnedSkills.newBuilder().setName(skill).setLevel(detail.skills.get(skill)));
			}

			protoPersonas.add(builder.setName(key)
					.setLevel(detail.level)
					.setArcana(Arcana.valueOf(detail.arcana.toUpperCase()))
					.build());
		}
		return protoPersonas;
	}
}

File: transformer/src/main/java/com/daykm/p5executioner/Main.kt
package com.daykm.p5executioner

import com.daykm.p5executioner.proto.*
import com.squareup.moshi.JsonAdapter
import com.squareup.moshi.Moshi
import com.squareup.moshi.Types
import okio.Okio
import java.io.File
import java.nio.file.Paths
import java.util.*

class JsonSpecialCombo(val result: String, val sources: List<String>)
class JsonCombo(val result: String, val source: Array<String>)
class JsonPersonaDetail(val arcana: String, val level: Int, val skills: Map<String, Int>, val elems: List<String>, val stats: Array<Int>)
class JsonSkillDetail(val element: String, val cost: Int?, val personas: Map<String, Int>, val talks: List<String>?, val effect: String)


val INPUT = "json/"

@Throws(Exception::class)
fun main(args: Array<String>) {
    println("Working Directory = " + System.getProperty("user.dir"))

    val data = with(Moshi.Builder().build()) {
        Data.newBuilder()
                .addAllArcanaCombos(createArcanaCombos(this))
                .addAllPersonas(createPersonas(this))
                .addAllSkills(createSkills(this))
                .addAllRareModifiers(createRareCombos(this))
                .addAllSpecialCombos(createSpecialCombos(this))
                .addAllDlcPersonae(createDlcPersonae(this))
                .build()
    }

    val bytes = data.toByteArray()

    println("Data is ${bytes.size} big")

    createOutputFile().outputStream().use { it.write(bytes) }
}

private fun createOutputFile(): File {

    val directoryPath = Paths.get("../gen/src/main/assets/")

    val directory = directoryPath.toFile().mkdirs()

    return Paths.get("../gen/src/main/assets/data.pb").toFile()
}

@Throws(Exception::class)
fun createDlcPersonae(moshi: Moshi): List<DLCPersona> = with(moshi) {
    adapter<List<List<String>>>(
            Types.newParameterizedType(List::class.java,
                    Types.newParameterizedType(List::class.java, String::class.java))
    ).fromJson(Okio.buffer(Okio.source(File(INPUT + "dlcPersonae.json"))))!!
            .mapTo(ArrayList<DLCPersona>()) {
                DLCPersona.newBuilder().addAllPersonaeInSeries(it).build()
            }
}

@Throws(Exception::class)
fun createSpecialCombos(moshi: Moshi): List<SpecialCombo> {
    val adapter = moshi.adapterForListOf(JsonSpecialCombo::class.java)

    val combos = adapter.fromJson(Okio.buffer(Okio.source(File(INPUT + "specialCombos.json"))))

    val protoCombos = ArrayList<SpecialCombo>(combos!!.size)

    combos.mapTo(protoCombos) { SpecialCombo.newBuilder().setResult(it.result).addAllSources(it.sources).build() }
    println("Parsed ${protoCombos.size} special combos")
    return protoCombos
}

@Throws(Exception::class)
fun createArcanaCombos(moshi: Moshi): List<ArcanaCombo> {
    val adapter = moshi.adapter<List<JsonCombo>>(Types.newParameterizedType(List::class.java, JsonCombo::class.java))
    val json = adapter.fromJson(Okio.buffer(Okio.source(File(INPUT + "arcanaCombos.json"))))!!
    return json.map {
        ArcanaCombo.newBuilder()
                .setFirst(Arcana.valueOf(it.source[0].toUpperCase()))
                .setSecond(Arcana.valueOf(it.source[1].toUpperCase()))
                .setResult(Arcana.valueOf(it.result.toUpperCase()))
                .build()
    }
}

@Throws(Exception::class)
fun createSkills(moshi: Moshi): List<Skill> {

    val type = Types.newParameterizedType(Map::class.java, String::class.java, JsonSkillDetail::class.java)

    val adapter = moshi.adapter<Map<String, JsonSkillDetail>>(type)

    val source = Okio.source(File(INPUT + "skills.json"))
    val skills = adapter.fromJson(Okio.buffer(source))!!

    val protoSkills = ArrayList<Skill>(skills.size)

    for (key in skills.keys) {
        val detail = skills[key]

        val builder = Skill.PersonaAndLevel.newBuilder()
        for (persona in detail!!.personas.keys) {
            builder.setPersona(persona).level = detail.personas[persona]!!
        }
        protoSkills.add(Skill.newBuilder()
                .addAllTalks(if (detail.talks != null) detail.talks else ArrayList<String>(0))
                .setCost(if (detail.cost != null) detail.cost else 0)
                .setName(key)
                .setEffect(detail.effect)
                .setElement(Skill.Element.valueOf(detail.element.toUpperCase()))
                .addPersonaeWhoLearn(builder)
                .build())
    }
    return protoSkills
}

@Throws(Exception::class)
fun createRareCombos(moshi: Moshi): List<RareComboModifier> {

    val type = Types.newParameterizedType(Map::class.java, String::class.java,
            Types.newParameterizedType(List::class.java, Int::class.javaObjectType))

    val adapter = moshi.adapter<Map<String, List<Int>>>(type)

    val source = Okio.source(File(INPUT + "rareCombos.json"))
    val rareCombos = adapter.fromJson(Okio.buffer(source))!!

    val protoModifiers = ArrayList<RareComboModifier>(rareCombos.size)

    rareCombos.keys.mapTo(protoModifiers) {
        RareComboModifier.newBuilder()
                .setArcana(Arcana.valueOf(it.toUpperCase()))
                .addAllModifiers(rareCombos[it])
                .build()
    }

    return protoModifiers
}

@Throws(Exception::class)
fun createPersonas(moshi: Moshi): List<Persona> {
    val adapter = moshi.adapterForMapOf(String::class.java, JsonPersonaDetail::class.java)

    val source = Okio.source(File(INPUT + "personae.json"))

    val personae = adapter.fromJson(Okio.buffer(source))!!

    val protoPersonas = ArrayList<Persona>(personae.size)

    for (key in personae.keys) {
        val detail = personae[key]
        val builder = Persona.newBuilder()

        val options = ArrayList<Persona.Affinities.AffinityOption>()

        for (affinity in detail?.elems!!) {
            when (affinity) {
                "ab" -> options.add(Persona.Affinities.AffinityOption.ABSORB)
                "-" -> options.add(Persona.Affinities.AffinityOption.NONE)
                "wk" -> options.add(Persona.Affinities.AffinityOption.WEAK)
                "rs" -> options.add(Persona.Affinities.AffinityOption.RESIST)
                "rp" -> options.add(Persona.Affinities.AffinityOption.REPEL)
                "nu" -> options.add(Persona.Affinities.AffinityOption.NULL)
                else -> options.add(Persona.Affinities.AffinityOption.UNRECOGNIZED)
            }
        }

        builder.setAffinities(Persona.Affinities.newBuilder()
                .setPhysical(options[0])
                .setGun(options[1])
                .setFire(options[2])
                .setIce(options[3])
                .setElectric(options[4])
                .setWind(options[5])
                .setPsychic(options[6])
                .setNuclear(options[7])
                .setBless(options[8])
                .setCurse(options[9]))

        builder.setStats(Persona.Stats.newBuilder()
                .setStrength(detail.stats[0])
                .setMagic(detail.stats[1])
                .setEndurance(detail.stats[2])
                .setAgility(detail.stats[3])
                .setLuck(detail.stats[4]))

        for (skill in detail.skills.keys) {
            builder.addSkills(
                    detail.skills[skill]?.let { Persona.LearnedSkills.newBuilder().setName(skill).setLevel(it) })
        }

        protoPersonas.add(builder.setName(key)
                .setLevel(detail.level)
                .setArcana(Arcana.valueOf(detail.arcana.toUpperCase()))
                .build())
    }
    return protoPersonas
}

fun <T> Moshi.adapterForListOf(clazz: Class<T>): JsonAdapter<List<T>> =
        adapter<List<T>>(Types.newParameterizedType(List::class.java, clazz))

fun <K, V> Moshi.adapterForMapOf(keyClass: Class<K>, valueClass: Class<V>): JsonAdapter<Map<K, V>> {
    val type = Types.newParameterizedType(Map::class.java, keyClass, valueClass)
    return adapter<Map<K, V>>(type)
}

